### DISCUSSION TITLE
Gallery of patterns?

### DISCUSSION THREAD

**Discussion by benhamilton** [jglev/obsidian-apply-patterns-plugin#43] [Ideas]
Could we have a gallery of pattern/command snippets? Thanks for this plugin, it's finally allowed me to do something I've wanted for a long time.

It took me some time to figure out how to get what I wanted, which was to be able to easily move tasks through a set of hashtags. I'm sure others would benefit, and the sharing of patterns might help others solve their particular problems too. 

I'll insert my pattern export below, which lets me move tasks through a specific set of statuses. Back [when I used Logseq, I asked for a feature to do essentially this](https://discuss.logseq.com/t/add-support-for-customisable-todo-keywords/696) which was based on [org-mode functionality](https://orgmode.org/guide/Multi_002dstate-Workflow.html). This solution, using **Apply Patterns** doesn't enforce the sequencing, but allows for the fast and easy transition of tasks through the set of hashtags to signify statuses that I use.

I've mapped ⌘⌃⌥⇧ to my CapsLock key ⇪ using BetterTouchTool on macOS, then I've mapped ⇪R to **Apply Patterns: Apply Pattern to whole lines** in Obsidian → Settings → Hotkeys

Which means that when on a line, or with a number of lines selected, I can now press ⇪R and get the following:
<img width="846" alt="CleanShot 2022-03-09 at 16 12 40@2x" src="https://user-images.githubusercontent.com/876530/157383181-3a884a7b-7457-4e92-939e-1ed210655089.png">

Which will mark the task appropriately and replace the hashtag with the tag I've asked for.

For example, tasks If I select #wip from the list, it'll transform 
`- [ ] this task here #todo` 
into
`- [/] this task here #wip`

or if I select #pending from the list, it'll transform
`- [/] this task here #wip`
into
`- [>] this task here #pending`

or if I select #done from the list, it'll transform 
`- [/] this task here #wip` 
into 
`- [x] this task here #done` 

and so forth for the various statuses.

Apply Patterns v2.0.0, pattern export as JSON:
```json
[
  {
    "name": "#backlog",
    "rules": [
      {
        "from": "- \\[.\\]",
        "to": "- [ ]",
        "caseInsensitive": false,
        "global": false,
        "multiline": false,
        "sticky": false,
        "disabled": false
      },
      {
        "from": "#(todo|inprogress|wip|pending|onhold|cancelled|done)",
        "to": "#backlog",
        "caseInsensitive": false,
        "global": false,
        "multiline": false,
        "sticky": false,
        "disabled": false
      }
    ],
    "collapsed": false,
    "cursorRegexStart": "$",
    "cursorRegexEnd": "^"
  },
  {
    "name": "#todo",
    "rules": [
      {
        "from": "- \\[.\\]",
        "to": "- [ ]",
        "caseInsensitive": false,
        "global": false,
        "multiline": false,
        "sticky": false,
        "disabled": false
      },
      {
        "from": "#(backlog|inprogress|wip|pending|onhold|cancelled|done)",
        "to": "#todo",
        "caseInsensitive": false,
        "global": false,
        "multiline": false,
        "sticky": false,
        "disabled": false
      }
    ],
    "collapsed": false,
    "cursorRegexStart": "$",
    "cursorRegexEnd": "^"
  },
  {
    "name": "#wip",
    "rules": [
      {
        "from": "- \\[.\\]",
        "to": "- [/]",
        "caseInsensitive": false,
        "global": false,
        "multiline": false,
        "sticky": false,
        "disabled": false
      },
      {
        "from": "#(backlog|todo|inprogress|pending|onhold|cancelled|done)",
        "to": "#wip",
        "caseInsensitive": false,
        "global": false,
        "multiline": false,
        "sticky": false,
        "disabled": false
      }
    ],
    "collapsed": false,
    "cursorRegexStart": "$",
    "cursorRegexEnd": "^"
  },
  {
    "name": "#pending",
    "rules": [
      {
        "from": "- \\[.\\]",
        "to": "- [>]",
        "caseInsensitive": false,
        "global": false,
        "multiline": false,
        "sticky": false,
        "disabled": false
      },
      {
        "from": "#(backlog|todo|inprogress|wip|onhold|cancelled|done)",
        "to": "#pending",
        "caseInsensitive": false,
        "global": false,
        "multiline": false,
        "sticky": false,
        "disabled": false
      }
    ],
    "collapsed": false,
    "cursorRegexStart": "$",
    "cursorRegexEnd": "^"
  },
  {
    "name": "#onhold",
    "rules": [
      {
        "from": "- \\[.\\]",
        "to": "- [>]",
        "caseInsensitive": false,
        "global": false,
        "multiline": false,
        "sticky": false,
        "disabled": false
      },
      {
        "from": "#(backlog|todo|inprogress|wip|onhold|cancelled|done)",
        "to": "#onhold",
        "caseInsensitive": false,
        "global": false,
        "multiline": false,
        "sticky": false,
        "disabled": false
      }
    ],
    "collapsed": false,
    "cursorRegexStart": "$",
    "cursorRegexEnd": "^"
  },
  {
    "name": "#cancelled",
    "rules": [
      {
        "from": "- \\[.\\]",
        "to": "- [-]",
        "caseInsensitive": false,
        "global": false,
        "multiline": false,
        "sticky": false,
        "disabled": false
      },
      {
        "from": "#(backlog|todo|inprogress|wip|pending|onhold|done)",
        "to": "#cancelled",
        "caseInsensitive": false,
        "global": false,
        "multiline": false,
        "sticky": false,
        "disabled": false
      }
    ],
    "collapsed": false,
    "cursorRegexStart": "$",
    "cursorRegexEnd": "^"
  },
  {
    "name": "#done",
    "rules": [
      {
        "from": "- \\[.\\]",
        "to": "- [x]",
        "caseInsensitive": false,
        "global": false,
        "multiline": false,
        "sticky": false,
        "disabled": false
      },
      {
        "from": "#(backlog|todo|wip|inprogress|pending|onhold|cancelled)",
        "to": "#done",
        "caseInsensitive": false,
        "global": false,
        "multiline": false,
        "sticky": false,
        "disabled": false
      }
    ],
    "collapsed": false,
    "cursorRegexStart": "$",
    "cursorRegexEnd": "^"
  }
]
```

The eagle-eyed will notice a status of `in progress` in the JSON above, which is a leftover vestige from previous attempts, but no longer use, and I've still got some of these around, so when I meet them, these patterns will convert those to the new status set I've configured.

**Comment by jglev** [comment]
@benhamilton, thank you for writing this up!

I agree that sharing patterns like this could be of use for other users. I actually wrote Apply Patterns _specifically_ to enable for myself the type of todo management you described above -- I also previously used, and have been heavily influenced by, org-mode's approach!

Sharing patterns is something that I would prefer not to write into the plugin itself, in favor of offloading it to some existing platform. I also will say that I'm not in a position currently to curate such a collection myself, though I'd welcome help with that!

With that in mind, I was first thinking that the [Obsidian forum](https://forum.obsidian.md/) might be a natural place to share snippets, similar to [this long-running](https://forum.obsidian.md/t/how-to-achieve-css-code-snippets/8474) stickied thread for CSS snippets. Alternatively, I also just now realized that GitHub's built-in Discussions "Category" feature might be well-suited to this -- So I just now [renamed the "Show and Tell" discussion category in this Repo to "Example Patterns"](https://github.com/jglev/obsidian-apply-patterns-plugin/discussions/categories/example-patterns).

Does that sort of approach -- a central discussion area centered in GitHub Discussions -- seem like it would be responsive to your idea? If so, I'd be happy to call it out / link to it in the Readme, so that it shows up when users are installing the plugin. I could also add a link in the Settings tab.

**Reply by benhamilton** [reply]
Hi Jacob, yes, I think that's a great idea. Putting examples that we all come up with into the "Example Patterns" (formerly Show and Tell) will work well, it also keeps the examples *close* to the plugin code, without being a burden on anyone. 
Thanks!

**Reply by jglev** [reply]
Great! And thank you, too, for discussing this!

**Comment by d9k** [comment]
Rule example to add checkbox (in my case `[-]` for "cancelled") for list item when no checkbox was present before:
Update: added one more case when list item starts as markdown link with title, like `- [Bing](https://bing.com)`

```json
    {
      "name": "checkbox: cancelled",
      "rules": [
        {
          "from": "- \\[.\\]",
          "to": "- [-]",
          "caseInsensitive": false,
          "global": false,
          "multiline": false,
          "sticky": false,
          "disabled": false
        },
        {
          "from": "^([ \\t]*)- ([^\\[])",
          "to": "$1- [-] $2",
          "caseInsensitive": false,
          "global": false,
          "multiline": false,
          "sticky": false,
          "disabled": false
        },
        {
          "from": "^([ \\t]*)- \\[(.[^\\]])",
          "to": "$1- [-] [$2",
          "caseInsensitive": false,
          "global": false,
          "multiline": false,
          "sticky": false,
          "disabled": false
        }
      ],
      "collapsed": false,
      "cursorRegexStart": "$",
      "cursorRegexEnd": "^"
    },
```
